ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt

```diff
@@ -0,0 +1,38 @@
+package kotlinx.coroutines.javafx
+
+import javafx.beans.value.ChangeListener
+import javafx.beans.value.ObservableValue
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.flowOn
+
+/**
+ * Creates an instance of a cold [Flow] that subscribes to the given [ObservableValue] and produces
+ * its values as they change.
+ *
+ * The resulting flow is conflated, meaning that if several values arrive in a quick succession, only
+ * the last one will be produced.
+ *
+ * It produces at least one value.
+ *
+ * Since this implementation uses [ObservableValue.addListener], even if this [ObservableValue]
+ * supports lazy evaluation, eager computation will be enforced while the flow is being collected.
+ */
+@ExperimentalCoroutinesApi
+public fun <T: Any> ObservableValue<T>.asFlow(): Flow<T> = callbackFlow<T> {
+    val listener = ChangeListener<T> { _, _, newValue ->
+        try {
+            offer(newValue)
+        } catch (e: CancellationException) {
+            // In case the event fires after the channel is closed
+        }
+    }
+    addListener(listener)
+    send(value)
```

Comment:
There is a problem with this behavior:
```kotlin
    @Test
    fun failure() = runTest {
        withContext(Dispatchers.JavaFx) {
            val END_MARKER = "xxx"
            val integerProperty = SimpleStringProperty()
            val flow = integerProperty.asFlow().takeWhile { j -> j != END_MARKER }
            launch {
                yield()
                yield()
                integerProperty.set("xxx")
            }
            flow.collect {

            }
        }
    }
```

Reply:
This test fails, because the initial value is `null`, and `null` can't be passed to `send` for some reason.

However, the problem is deeper:
```kotlin
    @Test
    fun failure() = runTest {
        withContext(Dispatchers.JavaFx) {
            val END_MARKER = "xxx"
            val integerProperty = SimpleStringProperty("xyzzy")
            val flow: Flow<String?> = integerProperty.asFlow().takeWhile {
                it != null && it == END_MARKER
            }
            launch {
                yield()
                integerProperty.set(null)
                yield()
                integerProperty.set(END_MARKER)
            }
            flow.collect {

            }
        }
    }
```
also fails, and this can't be easily fixed: https://docs.oracle.com/javase/8/javafx/api/javafx/beans/value/WritableIntegerValue.html#setValue-java.lang.Number- JavaFX is, as seen in many places, fine with storing `null` in properties.

Reply:
This test does not pass because the compiler inserts null-checks in `ChangeListener`.
And it does so because you have explicit `Any` bound: `public fun <T: Any> ObservableValue<T>.asFlow()`.

It looks like you can remove `: Any` upper bound and ignore the initial null value.

Reply:
Yes, that helped!